apps/pipeline-visualization/src/mermaid_builder.ts

```diff
@@ -0,0 +1,88 @@
+// SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+import { BlockDefinition } from '@jvalue/jayvee-language-server';
+
+export class MermaidBuilder {
+  protected mermaidLines: string[] = [];
+
+  pipelineHead(pipelineName: string): MermaidBuilder {
+    this.mermaidLines.push('subgraph ' + pipelineName);
+    return this;
+  }
+
+  blockType(block: BlockDefinition): MermaidBuilder {
+    this.mermaidLines.push(
+      `${block.name}[${block.name}<br><i>${block.type.ref?.name}</i>]`,
+    );
+    return this;
+  }
+
+  blockTypeWithProperties(
+    block: BlockDefinition,
+    propertyString: string,
+  ): MermaidBuilder {
+    this.mermaidLines.push(
+      `${block.name}[${block.name}<br><i>${block.type.ref?.name}</i><br>${propertyString}]`,
+    );
+    return this;
+  }
+
+  branching(block: BlockDefinition, child: BlockDefinition): MermaidBuilder {
+    this.mermaidLines.push(`${block.name}-->${child.name}`);
+    return this;
+  }
+
+  chain(chain: string[]): MermaidBuilder {
+    this.mermaidLines.push(chain.join('-->'));
+    return this;
+  }
+
+  compositeHeader(
+    block: BlockDefinition,
+    depth: number,
+    subgraphDirection: string,
+  ): MermaidBuilder {
+    this.mermaidLines.push(
+      `${'\t'.repeat(depth)}subgraph ${block.name}\n ${'\t'.repeat(
+        depth,
+      )}direction ${subgraphDirection}`,
+    );
+    return this;
+  }
+
+  compositeBody(subblocks: string[], depth: number): MermaidBuilder {
+    this.mermaidLines.push(`${'\t'.repeat(depth)}` + subblocks.join('-->'));
+    return this;
+  }
+
+  compositeTail(depth: number): MermaidBuilder {
+    this.mermaidLines.push(`${'\t'.repeat(depth)}end`);
+    return this;
+  }
+
+  classAssign(block: BlockDefinition, className: string): MermaidBuilder {
+    this.mermaidLines.push(`class ${block.name} ${className};`);
+    return this;
+  }
+
+  section(section: string): MermaidBuilder {
```

Comment:
What is a section?

Reply:
Ah, I got it when reading `mermaid_utils.ts`. Don't _build_ too early, just continue with the same builder (i.e. use just one instance of a builder = just one `new MermaidBuilder()`).

Reply:
remove sections and write lines directly